man/khepri.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "atum\-init" "8" 
.SH NAME
init \- a UNIX process no 1
.SH SYNOPSIS
\f[B]init\f[R] [ 0 | 6 ]
.SH DESCRIPTION
\f[B]atum\-init\f[R] is the first process the kernel starts.
If \f[B]atum\-init\f[R] is started as process no 1, it runs and
replaces itself with \f[B]atum\f[R](8).
.PP
If \f[B]atum\-init\f[R] is started while the system is up, it must be
either called as \f[B]init 0\f[R] or \f[B]init 6\f[R]:
.TP
\f[B]init 0\f[R]
tells the Unix process no 1 to shutdown and halt the system.
To signal \f[B]atum\f[R](8) the system halt request,
\f[B]atum\-init\f[R] removes all permissions of the file
\f[I]/etc/atum/reboot\f[R] (chmod 0), and sets the execute by owner
permission of the file \f[I]/etc/atum/stopit\f[R] (chmod 100).
Then a CONT signal is sent to \f[B]atum\f[R](8).
.TP
\f[B]init 6\f[R]
tells the Unix process no 1 to shutdown and reboot the system.
To signal \f[B]atum\f[R](8) the system reboot request,
\f[B]atum\-init\f[R] sets the execute by owner permission of the files
\f[I]/etc/atum/reboot\f[R] and \f[I]/etc/atum/stopit\f[R] (chmod 100).
Then a CONT signal is sent to \f[B]atum\f[R](8).
.SH EXIT CODES
\f[B]atum\-init\f[R] returns 111 on error, 0 in all other cases.
.SH SEE ALSO
atum(8), ennead(8), wepwawet(8), heka(8), nehebkau(8), khnum(8),
seshat(8)
.PP
index.html
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * khnum.c: -F, -I: establish signal handlers after fork.
  * seshat.c: zero-initialize allocated logdir array (thx github/ungerade);
    pick up processor outside SIGCHLD handler.
  * atum.h, atum.c, sig.h, sig.c, man/atum.8: add SIGPWR handling through
    /etc/atum/pwrfail (github/jchook).
  * khnum.c: -F, -I: don't fail if /proc does not exist.
  * doc/replaceinit.html: replace tee&&chmod with cat;chmod.
  * atum.h, nehebkau.c, ennead.c, Makefile: buffer size PATH_MAX instead of
    256 if NEHEBKAU_USE_SYSLIMITS is defined.
  * khnum.c, khnum.check, man/khnum.8: add new option -A: alarm(2) support
    (github/pilcrow).
  * heka.c: improve warning if ./log exists but not ./log/supervise/ok (thx
    github/Swiddis).
  * khnum.c, man/khnum.8: add new option -N: with -012 reopen to /dev/null;
    -2: dup() and coe() stderr, restore on failed exec (github/pilcrow).
  * man/ennead.8: typo.
  * doc/*.html: add color-scheme.
  * ennead.c: use inotify(7) (if available) to watch services directory;
    remove overcautious delay; cleanup removed services more quickly.
  * Makefile, TARGETS: add test for inotify_init() system call.
  * hasinotify.h1, hasinotify.h2, tryinotify.c: new.
  * ennead.c: add selfpipe; buffer and optimize readproctitle log; close
    unused filedescriptor.
  * khnum.c: -F, -I: move signal handler functions to another ifdef; do not
    relay SIGCHLD also in main pid.
  * heka.c: status: if log/supervise/ can't be accessed, don't print confusing
    additional line after warning message (introduced with 5fe1bc7).
  * etc/debian: rename to etc/sysv.
  * etc/freebsd: rename to etc/bsd.
  * etc/openbsd, etc/freebsd/getty-ttyv4, etc/debian/getty-tty5: remove;
    obsolete.
  * doc/replaceinit.html: revise sections "Replacing init on *BSD" and
    "Replacing sysvinit on GNU/Linux".
  * doc/install.html: how to install as non-root user; optionally use cc on
    *BSD.
  * man/khnum.8: document new -F and -I options.
  * khnum.c: -F option implies -I (thx github/andy-bower).
  * khnum.c: add new option -F: fghack using unshare(); khnum.c: add new
    option -I: run another init using unshare().
  * Makefile, TARGETS: add test for unshare() system call (_GNU_SOURCE).
  * hasunshare.h1, hasunshare.h2, tryunshare.c: new.
  * conf-cc: trim.
  * doc/install.html, doc/replaceinit.html, doc/runlevels.html,
    doc/useinit.html: strip '^# ' from code blocks for multiline copy&paste.
  * man/sv.8, doc/dependencies.html: rename check script to program.
  * doc/runscripts.html: clarify outdated status of the collection.
  * doc/*.html, man/*.8: replace; converted from markdown source.
  * seshat.c, man/seshat.8: -LL: print warning on data loss.
  * seshat.c, man/seshat.8: add option -L, lossy (thx github/teslamotors).
  * package/upgrade: remove utmpset program on upgrade.
  * doc/replaceinit.html: remove getty finish scripts that use utmpset;
    etc/debian/getty-tty5/finish, etc/freebsd/getty-ttyv4/finish,
    etc/openbsd/getty-ttyC4/finish: remove, obsolete.
  * doc/index.html, man/khepri.8, man/atum.8, man/nehebkau.8,
    man/ennead.8: remove reference to utmpset program.
  * utmpset.c, utmpset.check, utmpset.dist, man/utmpset.8, tryuwtmp.c,
    tryuwtmpx.c, uw_tmp.h?: remove, obsolete.
  * Makefile, package/commands, src/Makefile, TARGETS: remove utmpset and
    utmpx sysdep.

atum 2.3.1
Sat, 07 Feb 2026 11:11:01 +0100
  * doc/dependencies.html: wording; cruft.
  * nehebkau.c, man/nehebkau.8: run control/? with service process id as argument.
  * heka.c: always write to control pipe, do not look at the status.
  * seshat.c: fix uninitialized pointer read with -v and log directory
    mode r--.
  * Makefile, TARGETS, byte.h: add byte_zero to byte.a.
  * byte_zero.c: new.
  * nehebkau.c: omit stat, pid in ./supervise/ if NEHEBKAU_STATUS_BINARYONLY is
    defined.
  * khnum.dist, heka.dist, seshat.dist: adapt.
  * buffer_put.c: fix compilation warning: -Wempty-body (github/clan).
  * buffer_put.c, byte.h, byte_copy.c, byte_diff.c: fix compilation
    warnings: -Wdiscarded-qualifiers (github/clan).
  * heka.c: fix compilation warning: -Wdangling-else (github/clan).
  * byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c, byte_rchr.c, khnum.c,
    str_chr.c, str_len.c, heka.c, seshat.c: fix compilation warnings:
    -Wmisleading-indentation (github/clan).
  * TARGETS: remove prot.o (github/clan).
  * doc/benefits.html: spelling.
  * doc/faq.html: cleanup, spelling.
  * doc/usedietlibc.html: remove.
  * doc/index.html: remove references to dietlibc, spelling.
  * man/nehebkau.8, doc/nehebkau.8.html: sync; clarify wording wrt log service and
    custom control char o; style.
  * man/atum.8, man/sv.8, man/seshat.8: typo, groff.

atum 2.3.0
Sun, 09 Nov 2025 11:07:17 +0100
  * TARGETS: remove hasshsgr.h.
  * nehebkauctrl.*, nehebkaustat.*, svwaitdown.*, svwaitup.*: remove, obsolete;
    Makefile, TARGETS: remove nehebkauctrl, nehebkaustat, svwaitup, svwaitdown.
  * nehebkau.c: fix typo in pathname (thx github/plorenzo19).
  * heka.c, man/sv.8: fix "heka check" to behave as documented (thx Buck Evan,
    James Byrne).
  * hasflock.h1: add #include <unistd.h>.
  * doc/dependencies.html: rewrite; new initial terse version.
  * doc/index.html: replace broken link to sysvinit.
  * doc/runscripts.html: remove broken links and reference to mailing list.
  * doc/replaceinit.html: switch http links to https.
  * seshat.c: never prune the in-progress .u (thx Mike Pomraning for the
    draft patch).
  * heka.c: fix regression introduced with 5fe1bc7: exit code of heka status in
    "lsb" mode is not 3 in all cases when the service is "down" (thx Fabian
    Ruff, David Mountney for the patch).
  * wepwawet.c, man/wepwawet.8: the environment variable $ENNEAD
    now overrides the default working directory /etc/atum/ennead/
    (github/plorenzo19).
  * atum.c: really restart stage 2 if it crashed.
  * alloc.c, alloc_re.c, buffer_0.c, byte.h, byte_chr.c, byte_copy.c,

src/khepri.c
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <unistd.h>
#include "atum.h"
#include "strerr.h"
#include "sig.h"
#include "open.h"
#include "error.h"

#define USAGE " 0|6"
#define FATAL "init: fatal: "
/* #define WARNING "init: warning: " */

const char *progname;

void usage(void) { strerr_die4x(0, "usage: ", progname, USAGE, "\n"); }

void atum_halt () {
  if (open_trunc(STOPIT) == -1)
    strerr_die4sys(111, FATAL, "unable to create ", STOPIT, ": ");
  if (chmod(STOPIT, 0100) == -1)
    strerr_die4sys(111, FATAL, "unable to chmod ", STOPIT, ": ");
  if (chmod(REBOOT, 0) == -1)
    if (errno != error_noent)
      strerr_die4sys(111, FATAL, "unable to chmod ", REBOOT, ": ");
  kill(1, sig_cont);
  _exit(0);
}

void atum_reboot () {
  if (open_trunc(STOPIT) == -1)
    strerr_die4sys(111, FATAL, "unable to create ", STOPIT, ": ");
  if (chmod(STOPIT, 0100) == -1)
    strerr_die4sys(111, FATAL, "unable to chmod ", STOPIT, ": ");
  if (open_trunc(REBOOT) == -1)
    strerr_die4sys(111, FATAL, "unable to create ", REBOOT, ": ");
  if (chmod(REBOOT, 0100) == -1)
    strerr_die4sys(111, FATAL, "unable to chmod ", REBOOT, ": ");
  kill(1, sig_cont);
  _exit(0);
}

int main (int argc, const char * const *argv, char * const *envp) {
  const char *prog[2];

  progname =*argv++;

  if (getpid() == 1) {
    prog[1] =0;
    prog[0] ="atum";

    /* kernel is starting init, atum does the job. */
    execve(ATUM, (char *const *)prog, envp);

    /* serious error */
    strerr_die4sys(111, FATAL, "unable to start ", prog[0], ": ");
  }

  if (! *argv || ! **argv) usage();
  switch (**argv) {
  case '0':
    atum_halt();
    break;
  case '6':
    atum_reboot();
    break;
  case '-':
    if ((*argv)[1] == 'V')
      strerr_warn1("$Id$\n", 0);
  default:
    usage();
  }
  /* not reached */
  _exit(0);
}
